### PR TITLE
network: Reconnect to the same peers on startup

### PR DESCRIPTION
**Pull request by kortatu**
Issue #1396 
When a swarm node was restarted, kademlia connection table changed because the suggested peers were different. This incurred in a big traffic loas as the node had to push part of its chunks to the new peers.
We wanted to prioritize known peers (latest connected peers) when reconnecting. To do that, we are saving the current peers bzz addresses on shutdown, and load them in startup and try to connect them. Of course if we need more peers or some of them can't be contacted, the usual peer suggestion will take place.

### CODE REVIEW

File: network/hive.go [hunk 1]
```diff
@@ -249,13 +252,40 @@ func (h *Hive) loadPeers() error {
...
+func (h *Hive) suggestInitialPeers(conns []*BzzAddr) {
+	log.Info(fmt.Sprintf("%08x hive suggestInitialPeers() With %v saved connections", h.BaseAddr()[:4], len(conns)))
+	for _, addr := range conns {
+		log.Trace(fmt.Sprintf("%08x hive connect() suggested initial %08x", h.BaseAddr()[:4], addr.Address()[:4]))
+		under, err := enode.ParseV4(string(addr.Under()))
+		if err == nil {
+			log.Trace(fmt.Sprintf("%08x attempt to connect to bee %08x", h.BaseAddr()[:4], addr.Address()[:4]))
+			h.addPeer(under)
+		} else {
+			log.Warn(fmt.Sprintf("%08x unable to connect to bee %08x: invalid node URL: %v", h.BaseAddr()[:4], addr.Address()[:4], err))
```
Comment (zelig) [inline network/hive.go:280]:
```golang 
log.Warn("suggestInitialPeers: invalid enode ID", "self", hex.EncodeToString(h.BaseAddr()[:4],  "peer",  addr,    "err", err)
```      

not to use `fmt` in log

update:  I see now this is all over the original code, so TODO cleanup later.

File: network/hive.go [hunk 2]
```diff
@@ -249,13 +252,40 @@ func (h *Hive) loadPeers() error {
...
+	return errRegistering
+}
 
-	return h.Register(as...)
+func (h *Hive) suggestInitialPeers(conns []*BzzAddr) {
+	log.Info(fmt.Sprintf("%08x hive suggestInitialPeers() With %v saved connections", h.BaseAddr()[:4], len(conns)))
+	for _, addr := range conns {
+		log.Trace(fmt.Sprintf("%08x hive connect() suggested initial %08x", h.BaseAddr()[:4], addr.Address()[:4]))
+		under, err := enode.ParseV4(string(addr.Under()))
+		if err == nil {
```
Comment (zelig) [inline network/hive.go]: it is gostyle to write like this;

```golang
under, err := enode.ParseV4(string(addr.Under()))
		if err != nil {
                    log.Warn("suggestInitialPeers: invalid enode ID", "self", hex.EncodeToString(h.BaseAddr()[:4],  "peer",  addr,    "err", err)
                   continue
               }
               h.addPeer(under)
}
```

File: network/hive.go [hunk 3]
```diff
@@ -249,13 +252,40 @@ func (h *Hive) loadPeers() error {
...
+		}
+
+	} else {
+		go h.suggestInitialPeers(conns)
+	}
+	return errRegistering
+}
 
-	return h.Register(as...)
+func (h *Hive) suggestInitialPeers(conns []*BzzAddr) {
```
Comment (zelig) [inline network/hive.go]: this is not just suggestion like from the point of view of kademlia.
pls call it `connectPeers(conns  ...*BzzAddr)` and the argument

File: network/hive.go [hunk 4]
```diff
@@ -265,8 +295,22 @@ func (h *Hive) savePeers() error {
...
+			return true
+		}
+		log.Trace("saving connected peer", "OAddr", hexutil.Encode(p.OAddr), "UAddr", p.UAddr)
+		conns = append(conns, p.BzzAddr)
+		return true
+	})
 	if err := h.Store.Put("peers", peers); err != nil {
 		return fmt.Errorf("could not save peers: %v", err)
 	}
+
```
Comment (zelig) [inline network/hive.go:310]: for consistency, i would change line  308 to have "addrs" instead of "peers"

**Review by zelig — APPROVED** [review]
This looks very good. Minor comments only.
Question: do we always want this behaviour?

File: network/hive.go (hunk 4 above)
Comment (kortatu) [reply network/hive.go:310]: That will break backward compatibility. Are we sure we want that?

**Review by acud — APPROVED** [review]
good work @kortatu

File: network/hive.go [hunk 5]
```diff
@@ -265,8 +295,22 @@ func (h *Hive) savePeers() error {
 		peers = append(peers, pa)
 		return true
 	})
+
+	h.Kademlia.EachConn(nil, 256, func(p *Peer, i int) bool {
+		if p == nil {
```
[the remark this answers is not in the document]
Comment (kortatu) [reply network/hive.go]: Not sure, I copied from the EachAddress call. Maybe is not possible, although if there is no connection, the root of the pot will be nil. Maybe is just a sanity check.

[the remark this answers is not in the document]
Comment (kortatu) [reply network/hive.go]: I've just checked to call EachConn on an empty kademlia and nil is not returned. I also looked up all uses of EachConn and it seems we never check for nullity, so maybe I can get rid of that check.

**Comment by kortatu** [comment]
> This looks very good. Minor comments only.
> Question: do we always want this behaviour?

Well, that I don't know. It seems better in most situations, we save computation time and network traffic. Maybe when the swarm node has been off for a long period, those connected peers are less probable to be still online, but it may be worth trying...

**Comment by nolash** [comment]
@zelig @acud this was approved two weeks ago. Is there a reason why it's not being merged?

**Pull request merged**